nixos/modules/services/networking/teleport.nix

```diff
@@ -0,0 +1,95 @@
+{ config, pkgs, lib, ... }:
+
+with lib;
+
+let
+  cfg = config.services.teleport;
+  settingsYaml = pkgs.formats.yaml { };
+  pidFilePath = "/run/teleport.pid";
+  cmdlineArgs = [ "--pid-file=${pidFilePath}" ]
+    ++ optional cfg.debug.enable "--debug"
+    ++ optional cfg.diag.enable "--diag-addr=${cfg.diag.addr}:${toString cfg.diag.port}"
+    ++ optional (cfg.settings != { }) "--config=${settingsYaml.generate "teleport.yaml" cfg.settings}"
+    ++ cfg.extraOpts;
+in
+{
+  options = {
+    services.teleport = with lib.types; {
+      enable = mkEnableOption "Teleport service";
+
+      settings = mkOption {
+        type = settingsYaml.type;
+        default = { };
+        example = literalExpression ''
+          {
+            teleport = {
+              nodename = "client";
+              advertise_ip = "192.168.1.2";
+              auth_token = "60bdc117-8ff4-478d-95e4-9914597847eb";
+              auth_servers = [ "192.168.1.1:3025" ];
+          }
+        '';
+        description = ''
+          Contents of the <literal>teleport.yaml</literal> config file.
+          The <literal>--config</literal> arguments will only be passed if this set is not empty.
+
+          See <link xlink:href="https://goteleport.com/docs/setup/reference/config/"/>.
+        '';
+      };
+
+      debug.enable = mkEnableOption "debug";
+
+      diag = {
+        enable = mkEnableOption ''
+          endpoints for monitoring purposes. See <link xlink:href="https://goteleport.com/docs/setup/admin/troubleshooting/#troubleshooting/"/>
+        '';
+
+        addr = mkOption {
+          type = str;
+          default = "127.0.0.1";
+          description = "Metrics and diagnostics address.";
+        };
+
+        port = mkOption {
+          type = int;
+          default = 3000;
+          description = "Metrics and diagnostics port.";
+        };
+      };
+
+      extraOpts = mkOption {
+        type = listOf str;
+        default = [ ];
+        example = literalExpression ''
+          [
+            "--labels=Foo=Bar"
+          ]
+        '';
+        description = "Extra commandline options.";
+      };
+
+    };
+  };
+
+  config = mkIf config.services.teleport.enable {
+    environment.systemPackages = [ pkgs.teleport ];
+
+    systemd.services.teleport = {
+      wantedBy = [ "multi-user.target" ];
+      after = [ "network.target" ];
+      serviceConfig = {
+        ExecStart = "${pkgs.teleport}/bin/teleport start" +
+          optionalString (length cmdlineArgs != 0) (" \\\n  " +
+            concatMapStringsSep " \\\n  " builtins.toJSON cmdlineArgs);
```

Comment:
contrary to what we have implied earlier `toJSON` is *not* sufficient for escaping in systemd directives :frowning_face: see #154113. for time being it may be better to use `script`, `escapeShellArgs`, and exec the main process yourself instead of using `ExecStart`

Reply:
I'm going to replace this with: 
```nix
        ExecStart = ''${pkgs.teleport}/bin/teleport start \
          --pid-file=${pidFilePath} \
          ${optionalString cfg.debug.enable "--debug"} \
          ${optionalString cfg.diag.enable "--diag-addr=${cfg.diag.addr}:${toString cfg.diag.port}"} \
          ${optionalString (cfg.settings != { }) "--config=${settingsYaml.generate "teleport.yaml" cfg.settings}"} \
          ${cfg.extraOpts}
        '';
```